src/main/scala/com/snowflake/snowpark/internal/DataTypeParser.scala

```diff
@@ -0,0 +1,442 @@
+   *   Either a [[DecimalType]] or [[ParseError]].
+   */
+  private def parseDecimalType(
+      precisionStr: String,
+      scaleStr: Option[String],
+      original: String): Either[ParseError, DataType] = {
+    val scaleIfEmpty = Right(0)
+    for {
+      precision <- this.parseIntSafe(precisionStr, original)
+      scale <- scaleStr.fold[Either[ParseError, Int]](scaleIfEmpty)(this.parseIntSafe(_, original))
+    } yield DecimalType(precision, scale)
+  }
+
+  /**
+   * Parses an array type by recursively parsing its element type.
+   *
+   * @param elementType
+   *   String representation of the array's element type.
+   * @return
+   *   Either a [[ArrayType]] containing the parsed element type or a [[ParseError]].
+   */
+  private def parseArrayType(elementType: String): Either[ParseError, DataType] = {
+    this.parseDataType(elementType).map(ArrayType)
+  }
+
+  /**
+   * Parses a map type by splitting the inner types at the top-level delimiter and recursively
+   * parsing the key and value types.
+   *
+   * @param innerTypes
+   *   String representation of the map's inner key-value type specification.
+   * @param original
+   *   Original type string for error reporting.
+   * @return
+   *   Either a [[MapType]] containing the parsed key and value types or a [[ParseError]].
+   */
+  private def parseMapType(innerTypes: String, original: String): Either[ParseError, DataType] = {
+    this.splitAtTopLevelDelimiter(innerTypes, delimiter = ',') match {
+      case Some((keyStr, valueStr)) if keyStr.nonEmpty && valueStr.nonEmpty =>
+        for {
+          keyType <- this.parseDataType(keyStr)
+          valueType <- this.parseDataType(valueStr)
+        } yield MapType(keyType, valueType)
+      case _ =>
+        Left(MalformedType(original, expected = "map<key_type, value_type>"))
+    }
+  }
+
+  /**
+   * Normalizes input strings by converting to lowercase and trimming whitespace.
+   *
+   * @param input
+   *   The input string to normalize.
+   * @return
+   *   An [[Option]] containing the normalized string, or [[None]] if the input was null.
+   */
+  private def normalize(input: String): Option[String] =
+    Option(input).map(_.toLowerCase(Locale.ROOT).trim)
+
+  /**
+   * Safely parses a string to an integer with error handling.
+   *
+   * This method attempts to convert a string representation to an integer value, wrapping any
+   * parsing failures in a [[ParseError]] for consistent error handling throughout the parser.
+   *
+   * @param str
+   *   The string to parse as an integer.
+   * @param original
+   *   The original input string for error reporting context.
+   * @return
+   *   Either a [[ParseError]] if the string cannot be parsed as an integer, or the successfully
+   *   parsed integer value.
+   */
+  private def parseIntSafe(str: String, original: String): Either[ParseError, Int] =
+    Try(str.toInt).toEither.left.map(_ =>
+      InvalidInteger(original, s"'$str' is not a valid integer"))
```

Comment:
The error message being generated in the parseIntSafe function seems like it could be improved to avoid repeating the "invalid integer" idea twice.

For example, as it is right now, an error message looks like this:

```
Invalid integer '...': '...' is not a valid integer'
```

What do you think about changing it? It could look something like this:

```
Invalid integer '...': the value '...' could not be parsed
```

```suggestion
      InvalidInteger(original, s"'$str' could not be parsed"))
```